{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5ecfe42b",
   "metadata": {},
   "source": [
    "## Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5ea5d916-f5e4-4581-b0a8-c3d9596da094",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyBigWig\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.metrics import make_scorer\n",
    "from scipy.stats import spearmanr\n",
    "import xgboost as xgb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faaf799e",
   "metadata": {},
   "source": [
    "## Data Loading & Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69538a21-52df-42dd-a69c-9c1a115e5d42",
   "metadata": {},
   "source": [
    "We run big_wig_to_bw_renaming.sh to rename all data files to .bw format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81dea6c6-8d93-4479-b546-54fa14ebab0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The following script is available to execute as a .sh file to rename all the data files format to .bw\n",
    "#\n",
    "# #!/bin/bash\n",
    "#\n",
    "# # Check if the input variable $Path_to_ML4G_Project_1_Data is set\n",
    "# if [ -z \"$Path_to_ML4G_Project_1_Data\" ]; then\n",
    "#     echo 'Missing input variable $Path_to_ML4G_Project_1_Data. Please run big_wig_to_bw_renaming.sh with the correct path as an argument.'\n",
    "#     exit 1\n",
    "# fi\n",
    "#\n",
    "# # Navigate to the base data directory\n",
    "# cd \"$Path_to_ML4G_Project_1_Data\" || exit\n",
    "#\n",
    "# # Find and rename all bigWig files to .bw\n",
    "# find . -type f \\( -iname \"*.bigwig\" -o -iname \"*.bigWig\" -o -iname \"*.BigWig\" \\) -print0 | while IFS= read -r -d '' file; do\n",
    "#     # Get the directory and base name of the file\n",
    "#     dir=$(dirname \"$file\")\n",
    "#     base=$(basename \"$file\")\n",
    "#     # Construct the new file name with .bw extension\n",
    "#     newbase=\"${base%.*}.bw\"\n",
    "#     newfile=\"$dir/$newbase\"\n",
    "#     # Rename the file\n",
    "#     mv \"$file\" \"$newfile\"\n",
    "#     echo \"Renamed '$file' to '$newfile'\"\n",
    "# done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a5e553b4-3fce-48b0-b8da-912ad24bc786",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me1-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me3-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K27ac-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/DNase-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me1-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me3-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K27ac-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/DNase-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me1-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me3-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K27ac-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/DNase-bigwig/X1.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me1-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me3-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K27ac-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/DNase-bigwig/X2.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me1-bigwig/X3.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K4me3-bigwig/X3.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/H3K27ac-bigwig/X3.bw\n",
      "Loading this bigWig file: /Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/DNase-bigwig/X3.bw\n"
     ]
    }
   ],
   "source": [
    "# Load bigwig files and extract features for each gene \n",
    "def load_features(file_path, gene_df, window_length, num_bins):\n",
    "    \n",
    "    print(f\"Loading this bigWig file: {file_path}\")\n",
    "    bw = pyBigWig.open(file_path)\n",
    "    features = []\n",
    "    half = window_length // 2\n",
    "\n",
    "    for idx, row in gene_df.iterrows():\n",
    "        chromosome = row['chr'] \n",
    "        strand = row['strand']\n",
    "        tss = int(row['TSS_start'] if strand == '+' else row['TSS_end'])\n",
    "        begin = max(0, tss - half)  #to ensure not extending window to non-negative values\n",
    "        end = tss + half\n",
    "        bins = np.linspace(begin, end, num_bins + 1, dtype=int)\n",
    "        vals = []\n",
    "        for i in range(num_bins):\n",
    "            bin_begin = bins[i]\n",
    "            bin_end = bins[i + 1]\n",
    "          \n",
    "            bin_vals = bw.stats(chromosome, bin_begin, bin_end, type='mean')[0]\n",
    "            if bin_vals is None:\n",
    "                bin_vals = 0\n",
    "            vals.append(bin_vals)\n",
    "        features.append(vals)\n",
    "\n",
    "    bw.close()\n",
    "    features_df = pd.DataFrame(features)\n",
    "    features_df.columns = [f'bin_{i}' for i in range(num_bins)]\n",
    "    features_df['gene_name'] = gene_df['gene_name'].values\n",
    "    return features_df\n",
    "\n",
    "# Merge CAGE info and expression data \n",
    "def load_cage(info_path, expression_path=None):\n",
    "    \n",
    "    info_df = pd.read_csv(info_path, sep='\\t')\n",
    "    if expression_path:\n",
    "        expression_df = pd.read_csv(expression_path, sep='\\t')\n",
    "        merged_df = pd.merge(info_df, expression_df, on='gene_name')\n",
    "    else:\n",
    "        merged_df = info_df\n",
    "    return merged_df\n",
    "\n",
    "# Process a cell line with features \n",
    "def process_cell_line(line, hist_marks, window_length, num_bins, data, data_type='train'):\n",
    "    if data_type == 'train':\n",
    "        info_file = os.path.join(data, 'CAGE-train', 'CAGE-train', f'{line}_train_info.tsv')\n",
    "        expression_file = os.path.join(data, 'CAGE-train', 'CAGE-train', f'{line}_train_y.tsv')\n",
    "    elif data_type == 'validation':\n",
    "        info_file = os.path.join(data, 'CAGE-train', 'CAGE-train', f'{line}_val_info.tsv')\n",
    "        expression_file = os.path.join(data, 'CAGE-train', 'CAGE-train', f'{line}_val_y.tsv')\n",
    "    elif data_type == 'test':\n",
    "        info_file = os.path.join(data, 'CAGE-train', 'CAGE-train', f'{line}_test_info.tsv')\n",
    "        expression_file = None\n",
    "\n",
    "    gene_df = load_cage(info_file, expression_file)\n",
    "    final_df = gene_df.copy()\n",
    "\n",
    "    # Extract features for each histone mark \n",
    "    for mark in hist_marks:\n",
    "        if mark == 'DNase':\n",
    "            mark_df = 'DNase-bigwig'\n",
    "        else:\n",
    "            mark_df = f'{mark}-bigwig'\n",
    "\n",
    "        file_extension = '.bw'\n",
    "        bw_file = os.path.join(data, mark_df, f'{line}{file_extension}')\n",
    "\n",
    "        # Extract feature and include in df \n",
    "        features_df = load_features(bw_file, gene_df, window_length, num_bins)\n",
    "        feature_cols = [f'{mark}_bin_{i}' for i in range(num_bins)]\n",
    "        features_df.columns = feature_cols + ['gene_name']\n",
    "        final_df = pd.merge(final_df, features_df, on='gene_name')\n",
    "\n",
    "    return final_df\n",
    "\n",
    "\n",
    "# Define parameters: here 2000 and 20 determined to perform best \n",
    "window_length = 2000\n",
    "num_bins = 20           \n",
    "hist_marks = ['H3K4me1', 'H3K4me3', 'H3K27ac', 'DNase']\n",
    "data = '/Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/Project_1/ML4G_Project_1_Data/'\n",
    "\n",
    "# Process training data\n",
    "X1_train_data = process_cell_line('X1', hist_marks, window_length, num_bins, data, data_type='train')\n",
    "X2_train_data = process_cell_line('X2', hist_marks, window_length, num_bins, data, data_type='train')\n",
    "\n",
    "# Modify gene names so there is no overlap \n",
    "X1_train_data['gene_name'] = X1_train_data['gene_name'] + '_X1'\n",
    "X2_train_data['gene_name'] = X2_train_data['gene_name'] + '_X2'\n",
    "\n",
    "# training data\n",
    "X_train_full = pd.concat([X1_train_data, X2_train_data], ignore_index=True)\n",
    "\n",
    "if not os.path.exists('./Data'):\n",
    "    os.makedirs('./Data')\n",
    "    \n",
    "X_train_full.to_csv('./Data/X_train.tsv', sep='\\t', index=False)\n",
    "\n",
    "# validation data\n",
    "X1_val_data = process_cell_line('X1', hist_marks, window_length, num_bins,data, data_type='validation')\n",
    "X2_val_data = process_cell_line('X2', hist_marks, window_length, num_bins,data, data_type='validation')\n",
    "\n",
    "# Modify gene names again so there is no overlap \n",
    "X1_val_data['gene_name'] = X1_val_data['gene_name'] + '_X1'\n",
    "X2_val_data['gene_name'] = X2_val_data['gene_name'] + '_X2'\n",
    "\n",
    "# validation data\n",
    "X_validation = pd.concat([X1_val_data, X2_val_data], ignore_index=True)\n",
    "X_validation.to_csv('./Data/X_validation.tsv', sep='\\t', index=False)\n",
    "\n",
    "# test data (X3)\n",
    "X3_test_data = process_cell_line('X3', hist_marks, window_length, num_bins, data, data_type='test')\n",
    "X3_test_data.to_csv('./Data/X_3_test.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0444299f",
   "metadata": {},
   "source": [
    "## Model (XGBoost)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8a7cb00-f1e3-437c-9764-9c6edb3d6c9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting 3 folds for each of 48 candidates, totalling 144 fits\n",
      "Hyperparameters: {'colsample_bytree': 1.0, 'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 100, 'subsample': 1.0}\n",
      "Correlation on training set: 0.7795704957519383\n",
      "Correlation on validation set: 0.7888818601100172\n"
     ]
    }
   ],
   "source": [
    "# spearmann correlation \n",
    "def spearman_corr(y_true, y_pred):\n",
    "    corr, _ = spearmanr(y_true, y_pred)\n",
    "    return corr\n",
    "\n",
    "spearman_scorer = make_scorer(spearman_corr, greater_is_better=True)\n",
    "\n",
    "# Preapre training and validation data and exclude non-numeric columns \n",
    "X_train_df = pd.read_csv('./Data/X_train.tsv', sep='\\t')\n",
    "X_val_df = pd.read_csv('./Data/X_validation.tsv', sep='\\t')\n",
    "non_numeric = ['gene_name', 'gex', 'chr', 'strand']\n",
    "\n",
    "X_train = X_train_df.drop(columns=non_numeric)\n",
    "y_train = X_train_df['gex']\n",
    "X_val = X_val_df.drop(columns=non_numeric)\n",
    "y_val = X_val_df['gex']\n",
    "\n",
    "# Define paramteres needed for GridSearch with XGBoost model, fit the model and get best \n",
    "param_grid = {\n",
    "    'max_depth': [3, 5, 7],\n",
    "    'learning_rate': [0.1, 0.01],\n",
    "    'n_estimators': [100, 200],\n",
    "    'subsample': [0.8, 1.0],\n",
    "    'colsample_bytree': [0.8, 1.0],\n",
    "}\n",
    "\n",
    "xgb_model = xgb.XGBRegressor(\n",
    "    objective='reg:squarederror',\n",
    "    random_state=42\n",
    ")\n",
    "\n",
    "grid_search = GridSearchCV(\n",
    "    estimator=xgb_model,\n",
    "    param_grid=param_grid,\n",
    "    scoring=spearman_scorer,\n",
    "    cv=3,\n",
    "    n_jobs=-1,\n",
    "    verbose=1\n",
    ")\n",
    "\n",
    "grid_search.fit(X_train, y_train)\n",
    "best_model = grid_search.best_estimator_\n",
    "print(\"Hyperparameters:\", grid_search.best_params_)\n",
    "\n",
    "y_train_pred = best_model.predict(X_train)\n",
    "spearman_corr_train = spearmanr(y_train, y_train_pred)[0]\n",
    "print(f\"Correlation on training set: {spearman_corr_train}\")\n",
    "\n",
    "# Get spearman for validation set \n",
    "y_val_pred = best_model.predict(X_val)\n",
    "spearman_corr_val = spearmanr(y_val, y_val_pred)[0]\n",
    "print(f\"Correlation on validation set: {spearman_corr_val}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24aa463d",
   "metadata": {},
   "source": [
    "## Predicition on Test Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f61d5c4-7d19-4bd4-a908-4e7c904bcee3",
   "metadata": {},
   "outputs": [],
   "source": [
    "X3_test = pd.read_csv('./Data/X_3_test.tsv', sep='\\t')\n",
    "X3_features = X3_test.drop(columns=['gene_name', 'chr', 'strand'])\n",
    "pred = best_model.predict(X3_features)\n",
    "\n",
    "assert isinstance(pred, np.ndarray), 'Prediction array must be a numpy array'\n",
    "assert np.issubdtype(pred.dtype, np.number), 'Prediction array must be numeric'\n",
    "assert pred.shape[0] == len(X3_test), 'Each gene should have a unique predicted expression'\n",
    "\n",
    "save_dir = '/Users/gonuni/Desktop/College/CBB/3rd_Semester/ML4Genomics/Projects/ML4Genomics/project1/' \n",
    "file_name = 'gex_predicted.csv'        \n",
    "zip_name = \"Indilewitsch_Marie-Claire_Cardenal_Gonzalo_Project1.zip\" \n",
    "save_path = f'{save_dir}/{zip_name}'\n",
    "compression_options = dict(method=\"zip\", archive_name=file_name)\n",
    "\n",
    "test_genes  = pd.DataFrame({\n",
    "    'gene_name': X3_test['gene_name'],\n",
    "    'gex_predicted': pred.tolist()\n",
    "})\n",
    "test_genes.to_csv(save_path, compression=compression_options)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eaf88057-b766-4c7e-aee7-864ee5950320",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
